### DISCUSSION TITLE
Mimic luabind behavoiur

### DISCUSSION THREAD

**Discussion by DJYar** [neolithos/neolua#167] [Q&A]
Hello everyone!
I have some troubles implementing script engine inside game engine port from C++ to C#.
Original game uses luajit + luabind. There are scripts which are implementing UI, AI and some other entities.
Also it is possible to define own system, register callbacks and game engine will call them from its main loop (or when any specific event will occur).
I can't make changes in any of bundled scripts but can make some sort of preprocessor.

To export classes from native side I've copied binding example from [this Issue](https://github.com/neolithos/neolua/issues/102) and upgraded it a bit (using luabind sources as a reference). 
I had to make a helper class to be able to call either native or overloaded methods in script-registered class.
Luabind `class "" ()` syntax works somehow, also `super()` for constructor.

Main issue is to make calls to base native or scripted class. For example
```lua
class "TestClass" (NativeClass)

function TestClass:OverridedMethod()
    return -NativeClass:OverridedMethod()
end
```
or even
```lua
class "TestClass2" (TestClass)

function TestClass2:OverridedMethod()
    return -TestClass:OverridedMethod()
end
```
This tricky thing is made with upvalues in luabind but we have no upvalues in NeoLua so I have no idea how to mimic this behaivour.

[Current messy script engine prototype is here in gist](https://gist.github.com/DJYar/12c3ccb38b30a5f5953f2b08d818fafe)

Thanks in advance.

**Comment by neolithos** [comment]
Why no using `debug.getupvalue`?

But overall, I do not really understand the question. What code is generated with `class "" ()`?

**Reply by DJYar** [reply]
Hey, thank you for the answer. I'll try to describe luabind flow with my lame english (sorry about that 😢)

Function `class("TestDerived")` creates class representation object on the native side and creates globally visible table `TestDerived`. Also at this point luabind creates metatable for `TestDerived` table with object construction dispatcher (`__call`) and other utility things in it.

Then function returns another function (stage 2) which takes base types as an arguments. This stage2 function is copying base's metatables to class rep, updates global class registry and type cast graph and defines `super()` member that is visible only inside `__init()` function of derived class instance. Very common pattern is `function TestDerived:__init() super() --[[ ... ]] end` 

Also it is possible to call base member by its explicit type name. For example
```lua
function TestDerived:Sum(a, b)
    return BaseClass:Sum(a,b)
end
```

So it sounds like `super()` and all base classes are upvalues or I have to setfenv somehow.

I'm trying to port luabind class_rep/object_rep from [here](https://github.com/decimad/luabind-deboostified/blob/master/src/class_rep.cpp) but without class registry, cast graph, GC handling etc etc but I think that it will be impossible to do such complex thing without Lua VM api

**Reply by neolithos** [reply]
I'am also not a native speaker ;-).

The class initialization and the metatable stuff works fine within plain (Neo)Lua.

As I understand the `class_rep.cpp`. It create a new stack and code in the lua-machine. This is at this point not possible in NeoLua (compiled is compiled). But you can change the compile result by implementing your own `ILuaDebug`. This implementation can catch the `class("TestDerived")` call or the `__init` (to introduce a local variable `super`).

In this part you have all freedom to change everything.

**Reply by DJYar** [reply]
Sounds good, I'll look into it, thank you! I'll post my research results here later ;)
